### PR TITLE
Proposal for features in Numerics namespace

### PR DESCRIPTION
**Pull request by jppelteret**
Relates to #1502

### CODE REVIEW

**Comment by jppelteret** [comment]
@luca-heltai ping

**Comment by luca-heltai** [comment]
@jppelteret thanks! I'll watch this PR. 👍

**Comment by certik** [comment]
Thanks @jppelteret. Would it make sense to merge this with the [Expression](https://github.com/symengine/symengine/blob/4cc711a935277c0fa93cb97f4b8280a497e08e89/symengine/expression.h#L17) class?

**Comment by jppelteret** [comment]
I think that it may be possible. My biggest concern is that I need to have [the constructor](https://github.com/symengine/symengine/pull/1504/files#diff-0316f0f842cbc3244f74f085cde7bad9R105) for arithmetic types marked as explicit.  Would this be an issue? I notice that [the expression class](https://github.com/symengine/symengine/blob/4cc711a935277c0fa93cb97f4b8280a497e08e89/symengine/expression.h#L27-L42) does not do the same.

If that's no issue, then would you be amenable to me extending my current implementation to include the other constructors and operators present in the `Expression` class, and then replace the `Expression` class with a typedef to `Numerics::SymEngineNumber`?

**Comment by certik** [comment]
Explicit constructor is safer and recommended in C++, isn't it? @bluescarni do you remember the motivation why the constructor in `Expression` is not explicit?

@jppelteret is `SymEngineNumber` only for numbers? `Expression` is for any SymEngine expression. What's the advantage of using SymEngine for numbers only? Also isn't the derivative in (https://github.com/symengine/symengine/pull/1504/files#diff-6c1c18b825350733cca42d225b48b0b7R78) simply always equal to 0?

**Comment by jppelteret** [comment]
No, may represent any of the types that I have found to be sensible so far in the context of the FE method and constitutive modelling. So its got constructors that build a [single character named symbol](https://github.com/symengine/symengine/pull/1504/files#diff-0316f0f842cbc3244f74f085cde7bad9R134], [multicharacter named symbol or parsed expression](https://github.com/symengine/symengine/pull/1504/files#diff-0316f0f842cbc3244f74f085cde7bad9R150), and [function symbol](https://github.com/symengine/symengine/pull/1504/files#diff-0316f0f842cbc3244f74f085cde7bad9R159) as well as the numeric types.

**Comment by jppelteret** [comment]
> Also isn't the derivative in (https://github.com/symengine/symengine/pull/1504/files#diff-6c1c18b825350733cca42d225b48b0b7R78) simply always equal to 0?

Not necessarily. The `RCP` data member might be of symbolic type. Please see [this section of the unit test](https://github.com/symengine/symengine/pull/1504/files#diff-896c5f23caf9c857109f9ddadc518d3fR440) for examples that illustrate this functionality.

**Comment by jppelteret** [comment]
@certik For further context, here is [the best example within the current set of unit tests](https://github.com/symengine/symengine/pull/1504/files#diff-7ee7677247471e4fed8ed71d3f53985aR178) that really highlights how I wish to incorporate SymEngine into my workflow.

**Comment by certik** [comment]
@jppelteret are there some type of expressions that `SymEngineNumber` cannot contain? I don't yet fully understand the intended purpose.

**Comment by jppelteret** [comment]
> If `SymEngineNumber` can contain full symbolic expressions, then I would rename it to just `Expression` and merge the two classes.

Sorry, I didn't see this comment. I think that this would be a satisfactory solution. I think that there would be just a [few constructors](https://github.com/symengine/symengine/blob/4cc711a935277c0fa93cb97f4b8280a497e08e89/symengine/expression.h#L60-L80)  and [operators](https://github.com/symengine/symengine/blob/4cc711a935277c0fa93cb97f4b8280a497e08e89/symengine/expression.h#L84-L94) that I would need to implement in order to fully reproduce the existing interface of the `Expression` class. I'll look at doing this in the next few days.

**Comment by isuruf** [comment]
>  I think that there would be just a [few constructors](https://github.com/symengine/symengine/blob/4cc711a935277c0fa93cb97f4b8280a497e08e89/symengine/expression.h#L60-L80) and [operators](https://github.com/symengine/symengine/blob/4cc711a935277c0fa93cb97f4b8280a497e08e89/symengine/expression.h#L84-L94) that I would need to implement in order to fully reproduce the existing interface of the `Expression` class.

The implicit constructor with type `RCP<const Basic>` might be problematic. I tried removing it and there were a lot of code to fix. (See isuruf/symengine@c82b001a99e3aa5148c49b6623f70646722a2e0d)

**Comment by isuruf** [comment]
Fixed in https://github.com/symengine/symengine/pull/1507

**Comment by jppelteret** [comment]
I'm most of the way through an amendment to my proposal. I've got a few things to fix, but I'll push the changes in the mean time so that you can get this gist of how I've approached this.

**Comment by jppelteret** [comment]
@isuruf Please see https://github.com/symengine/symengine/pull/1504/commits/de172c8e09d9e0a0945a38ba38252b5a1ceef90c,  https://github.com/symengine/symengine/pull/1504/commits/50781bf8e15db8a6303b5ccdaf2680f45bc9804d  and https://github.com/symengine/symengine/pull/1504/commits/d9a1b143e8382dc55c6eca8ae7d1710bc8c2e902.

**Comment by jppelteret** [comment]
> The implicit constructor with type `RCP<const Basic>` might be problematic.

@isuruf I think that the constructor that takes floating point types should be marked as explicit, and not necessarily the one with an `RCP` or `Basic` as an argument. It's to avoid potential ambiguities when calling, for example, `std::sin(2)` which could be promoted to `std::sin(SymEngineNumber(2))` if the constructor is not explicit.

**Comment by isuruf** [comment]
```
@isuruf I think that the constructor that takes floating point types should be marked as explicit, and not necessarily the one with an `RCP` or `Basic` as an argument. It's to avoid potential ambiguities when calling, for example, `std::sin(2)` which could be promoted to `std::sin(SymEngineNumber(2))` if the constructor is not explicit.
```

We shouldn't overload a function like `std::sin`, instead we should have `SymEngine::sin(Expression)` and if we use `sin(2)` it'll use the one from `std` and `sin(Expression(2))` would use the one from SymEngine due to Argument-dependent lookup feature of C++.

**Comment by jppelteret** [comment]
I've added a mechanism to enable / disable the implicit and explicit constructors based off of a template parameter. That means that the existing `Expression` class can be typedef'd as a specialisation of this `Numerics::GenericExpression` (previously `Numerics::SymEngineNumber`) ,and should be able to be used as it was previously as the implicit constructor (from arithmetic types) is enabled. I have then given a typedef in the Numerics namespace to its counterpart that has only an explicit constructor (from arithmetic types).

**Comment by jppelteret** [comment]
All of the pre-existing tests pass locally, with only minor changes to `basic/test_basic.cpp` having been necessary.

**Comment by jppelteret** [comment]
FYI: You can see how the constructor specialisation works at https://godbolt.org/z/tsyEJx

File: symengine/numerics/math.h [hunk 1]
```diff
@@ -0,0 +1,732 @@
...
+
+} // namespace Numerics
+} // namespace SymEngine
+
+#ifndef DOXYGEN
+
+// Import math operations into standard namespace.
+// This gives one the ability to use them within more libraries with generic
+// templated number types.
+namespace std
```
Comment (isuruf) [inline symengine/numerics/math.h]: I'm not a fan of this. I think the `std` namespace should be left as is. Why is this necessary?

Comment (isuruf) [reply symengine/numerics/math.h]: For example, Eigen has the following about a scalar type `T`

> The math function should be defined in the same namespace than T, or in the std namespace though that second approach is not recommended.

Comment (jppelteret) [reply symengine/numerics/math.h]: There are other libraries that export their math functions to the `std` namespace. For example, [Trilinos Sacado](https://github.com/trilinos/Trilinos/blob/master/packages/sacado/src/new_design/Sacado_Fad_Exp_MathFunctions.hpp) does this, and its very convenient for reusing existing templated code (originally written with arithmetic types in mind) with these SymEngine numbers without further adaption.

We could do this import locally (as we have done [here](https://github.com/dealii/dealii/blob/master/include/deal.II/differentiation/ad/adolc_math.h) for ADOL-C numbers), but this is (a) annoying, (b) would likely be repeated by other libraries and therefore (c) runs the risk of having duplicate function definitions between cross-linked libraries, which would be difficult to resolve. 

I strongly believe that its a convenient and useful thing to do. To make this optional, one could guard this with a compile-time activated `#define`.

Comment (isuruf) [reply symengine/numerics/math.h]: > We could do this import locally (as we have done [here](https://github.com/dealii/dealii/blob/master/include/deal.II/differentiation/ad/adolc_math.h) for ADOL-C numbers), but this is (a) annoying, (b) would likely be repeated by other libraries and therefore (c) runs the risk of having duplicate function definitions between cross-linked libraries, which would be difficult to resolve.

I agree that doing this in dealii or other libraries is messy. I just don't understand where these are needed though. Do you have an example where these functions are used in dealii?

Comment (jppelteret) [reply symengine/numerics/math.h]: Here's [one example](https://github.com/dealii/dealii/blob/master/include/deal.II/base/symmetric_tensor.h#L2250-L2285) in out tensor library. There must be countless other examples in the code base, but that's the first obvious one that I could find. Most of our users would write boundary conditions and constitutive models with this basic syntax
```
template<typename T = double>
T func (T x){
  return std::sin(x)*std::cos(x);
}
```
and
```
template<typename T = double, int dim>
T neohookean_energy_function (SymmetricTensor<2/*rank*/,dim,T> C){
  return 0.5*(trace(C) - dim - std::log(determinant(C));
}
```
rather than
```
template<typename T = double>
T func (T x){
  using std::sin;
   using std::cos;
  return sin(x)*cos(x);
}
```
and
```
template<typename T = double, int dim>
T neohookean_energy_function (SymmetricTensor<2/*rank*/,dim,T> C){
  using std::log;
  return 0.5*(trace(C) - dim - log(determinant(C));
}
```

I would speculate that this is the same across most numerical codes that primarily think in terms of arithmetic types (both libraries and user codes). So importing these math functions into the `std` namespace would really be a huge help to those that want to use `SymEngine` as a "drop-in" replacement of sorts that is compatible with existing code bases.

Comment (isuruf) [reply symengine/numerics/math.h]: Okay, thanks for the example. Now I understand why an implicit constructor with int, float types would be bad, but I still don't like having two types (one with the implicit constructors and the one without)

How about declaring,
```
Expression std::sin(RCP<const Basic> &x)
```
instead of 
```
Expression std::sin(Expression &x)
```

This should fix the problem of conflicting with int and float types, right?

**Comment by bluescarni** [comment]
> Explicit constructor is safer and recommended in C++, isn't it? @bluescarni do you remember the motivation why the constructor in `Expression` is not explicit?

It's been a long time, but I would assume I did it like that just to be consistent with the rest of symengine (in terms of implicit construction, conversions, etc.). I tend to write only explicit constructors normally.

File: symengine/numerics/math.h (hunk 1 above)
Comment (jppelteret) [reply symengine/numerics/math.h]: I don't think that this would work in general. As far as I could tell, this wouldn't allow compound operations like `std::sin(std::sqrt(x))`. Or am I missing something?

Comment (bluescarni) [reply symengine/numerics/math.h]: Note that adding new overloads to ``std`` is not only bad practice, it is explicitly forbidden by the standard and technically it results in undefined behaviour (and yes, I am aware that it tends to work in practice).

Comment (jppelteret) [reply symengine/numerics/math.h]: Fair enough. If that's the case then I guess we'd consider another approach: We would then provide our users with a stand-alone header that does this and that they can optionally include themselves.

Comment (bluescarni) [reply symengine/numerics/math.h]: To be fair, non-operator customisation points are a sore point in C++, and the fact that people keep on feeling the impulse to put stuff in ``std`` is one of the symptoms of that :)

What I usually end up doing is to approach the problem in the opposite way: instead of trying to add new overloads/specialisations to ``std::`` functions (say, ``sin()``), I define a ``sin()`` function in one of my namespaces, provide implementations of ``sin(float)``/``sin(double)`` based on the ``std::`` functions and then allow the users to overload/specialise my implementation for their types via  ADL, partial template specialisation, overloading, etc., depending on the situation. There's no silver bullet though (or at least, I have not found it), and different solutions come with different tradeoffs in terms of complexity, efficiency, etc.

http://ericniebler.com/2014/10/21/customization-point-design-in-c11-and-beyond/
https://quuxplusone.github.io/blog/2018/03/19/customization-points-for-functions/

Comment (jppelteret) [reply symengine/numerics/math.h]: > it is explicitly forbidden by the standard and technically it results in undefined behaviour

Just for the record, this is exactly what is explained at https://en.cppreference.com/w/cpp/language/extending_std . Thanks for the info!

> then allow the users to overload/specialise my implementation for their types via ADL

Thanks as well for this tip. A couple of years ago I discussed the use of ADL versus overloads in `std` when starting to extend our library to support arbitrary number types. The problem that we ran into is that we have ~1 million line codebase, and doing this systematically as required was simply too burdensome. But I do agree that relying on ADL would be the most correct way to do this. Having all of the math operators in `SymEngine::Numerics` would of course permit this. 

In summary, I'd be OK with removing these imports from SymEngine and putting it into our own codebase.

Comment (isuruf) [reply symengine/numerics/math.h]: > I don't think that this would work in general. As far as I could tell, this wouldn't permit compound operations like `std::sin(std::sqrt(x))`. Or am I missing something?

It would because `Expression` is implicitly convertible to `RCP<const Basic>`

Comment (jppelteret) [reply symengine/numerics/math.h]: With the current implementation this is not the case:
```
    // Conversion to SymEngine types
    {
      const Expression x("x");
      const SE::RCP<SE::Basic> r1 (x);
      const SE::RCP<SE::Basic> r2 = x;
    }
```
leads to 
```
<path>/symengine-dev/symengine/tests/numerics/test_number_types.cpp:148:32: error: no viable conversion from 'const Expression' (aka 'const GenericExpression<false>') to 'const SE::RCP<SE::Basic>'
      const SE::RCP<SE::Basic> r2 = x;
                               ^    ~
<path>/symengine-dev/symengine/utilities/teuchos/Teuchos_RCPDecl.hpp:459:10: note: candidate constructor not viable: no known conversion from 'const Expression' (aka 'const GenericExpression<false>') to 'Teuchos::ENull' for 1st argument
  inline RCP(ENull null_arg = null);
         ^
<path>/symengine-dev/symengine/utilities/teuchos/Teuchos_RCPDecl.hpp:524:10: note: candidate constructor not viable: no known conversion from 'const Expression' (aka 'const GenericExpression<false>') to 'const RCP<SymEngine::Basic> &' for 1st argument
  inline RCP(const RCP<T>& r_ptr);
         ^
<path>/symengine-dev/symengine/utilities/teuchos/Teuchos_RCPDecl.hpp:540:10: note: candidate template ignored: could not match 'RCP' against 'GenericExpression'
  inline RCP(const RCP<T2>& r_ptr);
         ^
In file included from <path>/symengine-dev/symengine/tests/numerics/test_number_types.cpp:3:
<path>/symengine-dev/symengine/numerics/number_types.h:1313:22: error: implicit instantiation of undefined template 'SymEngine::Numerics::internal::Evaluator<Teuchos::RCP<SymEngine::Basic>, void>'
    return internal::Evaluator<ResultType>::value(get_RCP());
                     ^
<path>/symengine-dev/symengine/tests/numerics/test_number_types.cpp:147:36: note: in instantiation of function template specialization 'SymEngine::Numerics::GenericExpression<false>::operator RCP<Teuchos::RCP<SymEngine::Basic> >' requested here
      const SE::RCP<SE::Basic> r1 (x);
                                   ^
<path>/symengine-dev/symengine/numerics/utilities.h:288:8: note: template is declared here
struct Evaluator;
       ^
In file included from <path>/symengine-dev/symengine/tests/numerics/test_number_types.cpp:3:
<path>/symengine-dev/symengine/numerics/number_types.h:1313:43: error: incomplete definition of type 'SymEngine::Numerics::internal::Evaluator<Teuchos::RCP<SymEngine::Basic>, void>'
    return internal::Evaluator<ResultType>::value(get_RCP());
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
3 errors generated.
```
It might be possible to implement conversion through `operator=`, but I don't see how its possible to do this for the constructor `SE::RCP<SE::Basic> (Expression)`

Comment (jppelteret) [reply symengine/numerics/math.h]: > It would because `Expression` is implicitly convertible to `RCP<const Basic>`

Could you perhaps elaborate on how this was done in the current `SymEngine::Expression` class, so that I could replicate it?

Comment (isuruf) [reply symengine/numerics/math.h]: You are right. I only implemented it in https://github.com/symengine/symengine/pull/1507. See https://github.com/symengine/symengine/pull/1507/files#diff-5052f081551b80163e1096fe0a85fcefR134

Comment (jppelteret) [reply symengine/numerics/math.h]: Cool, thanks. I think that this might work. I need to remove all of the templates to `GenericExpression`, and will let you know the outcome once I've done that.

**Comment by isuruf** [comment]
I merged https://github.com/symengine/symengine/pull/1507. Sorry about the conflicts

File: symengine/numerics/utilities.h [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+}
+
+/**
+ * Robust conversion of a fundamental SymEngine reference counted pointer
+ * to its underlying type. This facilitates a change in context in which
+ * any general number is being used.
+ */
+template<typename SE_t, typename = typename std::enable_if<!std::is_same<SE_t,SE::Basic>::value>::type>
+SE::RCP<const SE::Basic>
+make_symengine_rcp(const SE::RCP<const SE_t> &ptr_number)
```
Comment (isuruf) [inline symengine/numerics/utilities.h]: Can you explain what you are trying to do here? Can't you just do
```
{
    return ptr_number;
}
```

File: symengine/numerics/utilities.h [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+        // Out "value" may be representing the result of
+        // a function (trig, exp, pow etc.).
+        // If evaluation fails, then either
+        // 1. We've asked SymEngine to evaluate a number type that it doesn't
+        // support, or
+        // 2. We have not fully resolved all symbols
+        ReturnType out = 0.0;
+
+        try {
+            out = static_cast<ReturnType>(eval_double(*value));
```
Comment (isuruf) [inline symengine/numerics/utilities.h]: Why not use this at the beginning and remove Lines 410-450?

**Comment by jppelteret** [comment]
Ok, so I've finally had a chance to rework my proposal. Trying to make the `Expression` class have only an explicit constructor for floats etc. was an unmitigated disaster. So instead of trying to work around this I've extended the `Expression` class with the basic features that I need, and have added only the essentials required to implement a `BatchOptimizer`. The current implementation is sufficiently flexible to allow other classes derived off of (or with the same interface as) the `Expression` class to be used in conjunction with the `BatchOptimizer`. I *think* that this should be satisfactory for everyone.

File: symengine/numerics/optimizer.h [hunk 4]
```diff
@@ -0,0 +1,1444 @@
...
+    /**
+     * A map type used to indicate which dependent variable is associated
+     * with which entry in the output vector.
+     *
+     * We use a custom comparator here because otherwise we can't use
+     * std::map::find; it is sensitive to the some data from the
+     * underlying SymEngine::Basic other than the value that it represents.
+     */
+    using map_dependent_expression_to_vector_entry_t
+        = std::map<const SymEngine::Basic *const, std::size_t,
```
Comment (isuruf) [inline symengine/numerics/optimizer.h]: Instead of storing raw pointers, use RCP here for reference counting

Comment (jppelteret) [reply symengine/numerics/optimizer.h]: I believe that I tried this at first, but it didn't work correctly in this context:
```c++
const auto &it = map_dep_ptr_basic_vec_entry.find(&*(func.get_basic()));
```
As best as I can recall, if one uses `RCP<Basic>` then `find` tries to match the `RCP` and not the underlying `Basic`.  However, I'll try again to switch this over to using `RCP<Basic>` and see if I can make it work.

Comment (jppelteret) [reply symengine/numerics/optimizer.h]: I think that the missing element when I first implemented this was to use `RCPBasicKeyLess` as the comparator. I've made the change to `
std::map<RCP<const Basic>, std::size_t, RCPBasicKeyLess>` and I think that this now works as expected. Thanks for the suggestion!

**Comment by jppelteret** [comment]
I think that this PR is ready for some more thorough scrutiny. I'd be happy to break it up into a few smaller PR's, if that would make reviewing it easier.

**Comment by certik** [comment]
I think it looks good in general, it adds a new feature, but doesn't slow current things down.

It would be nice to have a jupyter notebook demo for this new feature.

@isuruf, can you please review some of the details?

**Comment by isuruf** [comment]
I'll have a look at the code soon. One issue I have is that the `evaluate` function has overheads other than just evaluating. I'd like evaluate to have no overhead other than calling the visitors' `call` functions.

**Comment by isuruf** [comment]
Btw, what's the difference between `substitute` and `evaluate`. I forgot the details about this.

**Comment by isuruf** [comment]
Also, what's the usage of `ExpressionType` template parameter in `BatchOptimizer` do? Is there a usecase other than `ExpressionType=Expression`?

**Comment by jppelteret** [comment]
> what's the usage of ExpressionType template parameter in BatchOptimizer? Is there a usecase other than ExpressionType=Expression

Yes, this centres around the use of `NumberWrapper`. I'm working on a testcase to demonstrate this, but its been so long since I implemented this in my own work that it'll take me some time to develop a minimal problem to add to the testsuite.

**Comment by jppelteret** [comment]
> what's the difference between substitute and evaluate

I believe that `substitute` allows the (partial) substitution of symbols expressions, but does not necessarily render a primitive at the end. `Evaluate` converts some result to a numeric primitive (or user-defined one, which is what needs to be robustly accommodated). I'm not sure if `evaluate` does substitution as well (I don't think so), which is why I call the former before the latter.

**Comment by jppelteret** [comment]
> what's the usage of ExpressionType template parameter in BatchOptimizer? Is there a usecase other than ExpressionType=Expression?

Sorry its taken me some time to get to this. I've now added an example that demonstrates in some manner why one would want the `ExpressionType` to be templated in the `DictionarySubstitutionVisitor`:

https://github.com/symengine/symengine/pull/1504/files#diff-7ee7677247471e4fed8ed71d3f53985aR381

The actual context is maybe not so relevant (see #1517) but it still highlights that custom number types demand it. I think that the `ExpressionType` template parameter for `BatchOptimizer` could be avoided by baking it directly into these structs:

https://github.com/symengine/symengine/pull/1504/files#diff-3c7b4aaffa260cf056b5d2153425f33bR202

I thought I'd first mention the possibility and open a discussion before changing the code.

**Comment by jppelteret** [comment]
@isuruf Is there anything that you're wanting me to do here?

> One issue I have is that the evaluate function has overheads other than just evaluating. I'd like evaluate to have no overhead other than calling the visitors' call functions.

Were you referring to [this function](https://github.com/symengine/symengine/pull/1504/files#diff-3c7b4aaffa260cf056b5d2153425f33bR731)? If so, then there's [this alternative](https://github.com/symengine/symengine/pull/1504/files#diff-3c7b4aaffa260cf056b5d2153425f33bR722) that returns all of the evaluated results at once. I need the former in order to do some management when there are a number of contextually unrelated expression that are simultaneously optimised.

**Comment by jppelteret** [comment]
I think that I'm going to refactor the `BatchOptimizer` class to make it more compact and more aligned with the other SymEngine class structures. I'll open another PR later when I'm done.

**Comment by certik** [comment]
@jppelteret thanks. Do you need some input from @isuruf on this?

**Comment by jppelteret** [comment]
I don't think so thanks. I've got an idea as to how I wish to take this forward.
